smtp/client.js

```diff
@@ -68,66 +121,89 @@ class Client {
 		this.smtp.connect(connect);
 	}
 
+	/**
+	 * @param {Message|MessageStack} msg msg
+	 * @param {function(Error, MessageStack): void} callback callback
+	 * @returns {void}
+	 */
 	send(msg, callback) {
-		if (
-			!(msg instanceof message.Message) &&
-			msg.from &&
-			(msg.to || msg.cc || msg.bcc) &&
-			(msg.text !== undefined || this._containsInlinedHtml(msg.attachment))
-		) {
-			msg = message.create(msg);
+		/**
+		 * @param {MessageStack} m message stack
+		 * @returns {boolean} can make message
+		 */
+		const canMakeMessage = m => {
+			return (
+				m.from &&
+				(m.to || m.cc || m.bcc) &&
+				(m.text !== undefined || this._containsInlinedHtml(m.attachment))
+			);
+		};
+
+		/**
+		 * @type {Message}
+		 */
+		const message =
+			msg instanceof Message ? msg : canMakeMessage(msg) ? create(msg) : null;
+
+		if (message == null) {
+			// prettier-ignore
+			callback(new Error('message is not a valid Message instance'), /**@type {MessageStack}*/(msg));
+			return;
 		}
 
-		if (msg instanceof message.Message) {
-			msg.valid((valid, why) => {
-				if (valid) {
-					const stack = {
-						message: msg,
-						to: addressparser(msg.header.to),
-						from: addressparser(msg.header.from)[0].address,
-						callback: (callback || function() {}).bind(this),
-					};
-
-					if (msg.header.cc) {
-						stack.to = stack.to.concat(addressparser(msg.header.cc));
-					}
+		message.valid((valid, why) => {
+			if (valid) {
+				const stack = {
+					message,
+					to: addressparser(message.header.to),
+					from: addressparser(message.header.from)[0].address,
+					callback: (callback || function() {}).bind(this),
+				};
 
-					if (msg.header.bcc) {
-						stack.to = stack.to.concat(addressparser(msg.header.bcc));
-					}
+				if (message.header.cc) {
+					stack.to = stack.to.concat(addressparser(message.header.cc));
+				}
 
-					if (
-						msg.header['return-path'] &&
-						addressparser(msg.header['return-path']).length
-					) {
-						stack.returnPath = addressparser(
-							msg.header['return-path']
-						)[0].address;
-					}
+				if (message.header.bcc) {
+					stack.to = stack.to.concat(addressparser(message.header.bcc));
+				}
 
-					this.queue.push(stack);
-					this._poll();
-				} else {
-					callback(new Error(why), msg);
+				if (
+					message.header['return-path'] &&
+					addressparser(message.header['return-path']).length
+				) {
+					stack.returnPath = addressparser(
+						message.header['return-path']
+					)[0].address;
 				}
-			});
-		} else {
-			callback(new Error('message is not a valid Message instance'), msg);
-		}
+
+				this.queue.push(stack);
+				this._poll();
+			} else {
+				// prettier-ignore
+				callback(new Error(why), /**@type {MessageStack}*/(msg));
+			}
+		});
 	}
 
+	/**
+	 * @param {*} attachment attachment
+	 * @returns {boolean} does contain
+	 */
 	_containsInlinedHtml(attachment) {
 		if (Array.isArray(attachment)) {
-			return attachment.some(() => {
```

Comment:
this flat-out did not look right to me; i'm almost certain it's a bug.

Reply:
+1 is there a test that could have failed here?

Reply:
[this one](https://github.com/zackschuster/emailjs/blob/drop-starttls/test/message.js#L277), potentially